cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql

```diff
@@ -0,0 +1,62 @@
+/**
+ * @name Errors When Using Variable Declaration Inside Loop
+ * @description Using variables with the same name is dangerous.
+ *              However, such a situation inside the while loop can lead to a violation of the accessibility of the program.
+ *              Requires the attention of developers.
+ * @kind problem
+ * @id cpp/errors-when-using-variable-declaration-inside-loop
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-1126
+ */
+
+import cpp
+
+/**
+ * Errors when using a variable declaration inside a loop.
+ */
+class DangerousWhileLoop extends WhileStmt {
+  Expr exp;
+  Declaration dl;
+
+  DangerousWhileLoop() {
+    this = dl.getParentScope().(BlockStmt).getParent*() and
+    exp = this.getCondition().getAChild*() and
+    not exp instanceof PointerFieldAccess and
+    not exp instanceof ValueFieldAccess and
+    exp.toString() = dl.getName() and
```

Comment:
Please don't use `toString` in query logic.  The behaviour of `toString` can be unpredictable and may change in future without consideration for uses like this.  You _can_ use `getName` to compare the names of things (variables, functions etc) as `getName` will always return the name.

Are we looking for a `VariableAccess va` with `va.getTarget().getName() = dl.getName()`?

Can we then get rid of some of the exceptions (lines beginning with `not`)?

Reply:
thanks for that suggestion.
I used it partially.
unfortunately I could not remove additional conditions.